app/models/miq_server/worker_management/monitor/kubernetes.rb

```diff
@@ -0,0 +1,100 @@
+module MiqServer::WorkerManagement::Monitor::Kubernetes
+  extend ActiveSupport::Concern
+  attr_accessor :pod_resource_version
+
+  def current_pods
+    @current_pods ||= Concurrent::Hash.new
+  end
+
+  def reset_current_pods
+    @current_pods = nil
+    current_pods
+  end
+
+  def cleanup_failed_deployments
+    ensure_pod_monitor_started
+
+    # TODO: We should have a list of worker deployments we'll delete to avoid accidentally killing pg/memcached/orchestrator
+    # See ContainerOrchestrator#get_pods
+    failed_deployments.each do |failed|
+      orchestrator.delete_deployment(failed)
+    end
+  end
+
+  def failed_deployments(restart_count = 5)
+    # TODO: This logic might flag deployments that are hitting memory/cpu limits or otherwise not really 'failed'
+    current_pods.select { |_name, h| h.fetch(:last_state_terminated) && h.fetch(:container_restarts, 0) > restart_count }.collect { |_name, h| h[:label_name] }
+  end
+
+  private
+  def start_pod_monitor
+    @monitor_thread ||= begin
+      reset_current_pods
+      Thread.new { monitor_pods }
+    end
+  end
+
+  def ensure_pod_monitor_started
+    if @monitor_thread.nil? || !@monitor_thread.alive?
+      if !@monitor_thread.nil? && @monitor_thread.status.nil?
+        dead_thread, @monitor_thread = @monitor_thread, nil
+        _log.info("#{log_prefix} Waiting for the Monitor Thread to exit...")
+        dead_thread.join
+      end
+
+      start_pod_monitor
+    end
+  end
+
+  def orchestrator
+    @orchestrator ||= ContainerOrchestrator.new
+  end
+
+  def monitor_pods
+    # TODO: To ensure we're in sync, we might want to break out of the watch, reset the current_pods and run this again
+    collect_initial_pods
+    loop { watch_for_pod_events }
+  end
+
+  def collect_initial_pods
+    pods = orchestrator.get_pods
+    pods.each { |p| save_pod(p) }
+    self.pod_resource_version = pods.resourceVersion || 0
+  end
+
+  def watch_for_pod_events
+    orchestrator.watch_pods(self.pod_resource_version) do |event|
+      case event.type.downcase
+      when "added", "modified"
+        save_pod(event.object)
+      when "deleted"
+        delete_pod(event.object)
+      when "error"
+        # We sometimes get error 'type' watch events with 410 codes in the status object
+        # https://github.com/fabric8io/fluent-plugin-kubernetes_metadata_filter/pull/243/files
+        break if event.object && event.object.code == 410
+      when "status"
+        # other times, we can get 'status' type/kind, with a code of 410
+        # https://github.com/ManageIQ/manageiq-providers-kubernetes/blob/745ba1332fa43cfb0795644279f3f55b8751f1c8/app/models/manageiq/providers/kubernetes/container_manager/refresh_worker/watch_thread.rb#L48
+        break if event.code == 410
```

Comment:
@agrare I couldn't track down why we're doing one thing in the kubernetes provider here and something different in the fluent-plugin. referenced above... maybe different api versions return a toplevel status object which can be used here and other api versions return an error watch event with the status object inside?  I'll need to track this down either way.

Reply:
I don't know that k8s defines the event type when a 410 Gone is returned, here are the docs on watches https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes so this might be an implementation details (whether it returns "status" or "error" as the event type).

Reply:
yeah, at the worst, if we have to support both formats for 410, we can combine them like:

```ruby
      when "error", "status"
        break if event.code == 410                        # outer watch event has failed 410 error code
        break if event.object && event.object.code == 410 # or outer watch event completed but inner event object has the 410 error code
```

It's less than ideal but if the outer watch event contains the code or the inner object contains the code,

Reply:
I suspect depending on what you are asking for, you can get "success" but with a failure object OR the whole request can fail.  It's possible there are bugs on the kubernetes side in terms of consistency in various API while there are possibly legitimate reasons to have a successful failure vs. a failing failure.